Adjustments in dataset

In [1]:
import pandas as pd

In [2]:
pd.set_option('display.max_colwidth', 0)
input_file = "../datasets/properties.json"
output_file = "../datasets/properties.csv"

In [3]:
df = pd.read_json(input_file, lines=True, dtype=False)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2772 entries, 0 to 2771
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   data    2772 non-null   object
dtypes: object(1)
memory usage: 21.8+ KB


In [4]:
df = pd.json_normalize(df.data)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2772 entries, 0 to 2771
Data columns (total 26 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   link                              2772 non-null   object
 1   title                             2772 non-null   object
 2   dt_scraping                       2772 non-null   object
 3   key_1                             2772 non-null   object
 4   area_util                         2772 non-null   object
 5   valor_rs_m2                       2772 non-null   object
 6   bairro                            2772 non-null   object
 7   cidade                            2772 non-null   object
 8   quartos                           2771 non-null   object
 9   suites                            1268 non-null   object
 10  garagens                          1738 non-null   object
 11  condominio_rs                     2197 non-null   object
 12  codigo              

Deleting unnecessary columns

In [5]:
df.drop(columns=["codigo", 
                 "ultima_atualizacao", 
                 "nome_do_edificio", 
                 "total_de_andar_do_empreendimento", 
                 "posicao_do_sol",
                 "posicao_do_imovel"],
        inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2772 entries, 0 to 2771
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   link                  2772 non-null   object
 1   title                 2772 non-null   object
 2   dt_scraping           2772 non-null   object
 3   key_1                 2772 non-null   object
 4   area_util             2772 non-null   object
 5   valor_rs_m2           2772 non-null   object
 6   bairro                2772 non-null   object
 7   cidade                2772 non-null   object
 8   quartos               2771 non-null   object
 9   suites                1268 non-null   object
 10  garagens              1738 non-null   object
 11  condominio_rs         2197 non-null   object
 12  area_total            1397 non-null   object
 13  iptu_rs               1429 non-null   object
 14  andar_do_apartamento  630 non-null    object
 15  unidades_no_andar     229 non-null    

Renaming column names

In [6]:
df.rename(columns={"key_1"                : "rental_value",
                   "area_util"            : "useful_area",
                   "valor_rs_m2"          : "m2_value",
                   "bairro"               : "neighborhood",
                   "cidade"               : "city",
                   "quartos"              : "rooms",
                   "garagens"             : "parking",
                   "condominio_rs"        : "condo_value",
                   "suites"               : "suites",
                   "area_total"           : "total_area",
                   "iptu_rs"              : "iptu",
                   "andar_do_apartamento" : "floor",
                   "unidades_no_andar"    : "floor_properties",
                   "aceita_financiamento" : "financing",
                   "aceita_permuta"       : "exchange",
                   "agio_rs"              : "agio",
                   "area_terreno"         : "land_area"}, 
            inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2772 entries, 0 to 2771
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   link              2772 non-null   object
 1   title             2772 non-null   object
 2   dt_scraping       2772 non-null   object
 3   rental_value      2772 non-null   object
 4   useful_area       2772 non-null   object
 5   m2_value          2772 non-null   object
 6   neighborhood      2772 non-null   object
 7   city              2772 non-null   object
 8   rooms             2771 non-null   object
 9   suites            1268 non-null   object
 10  parking           1738 non-null   object
 11  condo_value       2197 non-null   object
 12  total_area        1397 non-null   object
 13  iptu              1429 non-null   object
 14  floor             630 non-null    object
 15  floor_properties  229 non-null    object
 16  land_area         45 non-null     object
 17  financing     

Change column types to number types.
Due to pandas convertion float NaN to integer:
1) For int types, use fillna method before transforming
2) For float types, use fillna method after transforming

In [7]:
df.rental_value = df.rental_value.str.replace(".","",regex=True)
df.rental_value = df.rental_value.astype(float)
df.rental_value.fillna(value=0, inplace=True)

df.useful_area = df.useful_area.str.replace(",",".",regex=True)
df.useful_area = df.useful_area.str.replace("[ m²]","",regex=True)
df.useful_area = df.useful_area.astype(float)
df.useful_area.fillna(value=0, inplace=True)

df.m2_value = df.m2_value.astype(float)
df.m2_value.fillna(value=0, inplace=True)

df.rooms.fillna(value=0, inplace=True)
df.rooms = df.rooms.astype(int)

df.parking.fillna(value=0, inplace=True)
df.parking = df.parking.astype(int)

df.condo_value = df.condo_value.str.replace(".","",regex=True)
df.condo_value = df.condo_value.str.replace(",",".",regex=True)
df.condo_value = df.condo_value.astype(float)
df.condo_value.fillna(value=0, inplace=True)

df.total_area = df.total_area.str.replace(".","",regex=True)
df.total_area = df.total_area.str.replace(",",".",regex=True)
df.total_area = df.total_area.str.replace("[ m²]","",regex=True)
df.total_area = df.total_area.astype(float)
df.total_area.fillna(value=0, inplace=True)

df.iptu = df.iptu.str.replace(".","",regex=True)
df.iptu = df.iptu.str.replace(",",".",regex=True)
df.iptu = df.iptu.astype(float)
df.iptu.fillna(value=0, inplace=True)

df.suites.fillna(value=0, inplace=True)
df.suites = df.suites.astype(int)

df.floor.fillna(value=0, inplace=True)
df.floor = df.floor.str.replace("\D", "", regex=True)
df.floor = df.floor.fillna(value=0)
df.floor = df.floor.astype(int)

df.fillna(value="", inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2772 entries, 0 to 2771
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   link              2772 non-null   object 
 1   title             2772 non-null   object 
 2   dt_scraping       2772 non-null   object 
 3   rental_value      2772 non-null   float64
 4   useful_area       2772 non-null   float64
 5   m2_value          2772 non-null   float64
 6   neighborhood      2772 non-null   object 
 7   city              2772 non-null   object 
 8   rooms             2772 non-null   int32  
 9   suites            2772 non-null   int32  
 10  parking           2772 non-null   int32  
 11  condo_value       2772 non-null   float64
 12  total_area        2772 non-null   float64
 13  iptu              2772 non-null   float64
 14  floor             2772 non-null   int32  
 15  floor_properties  2772 non-null   object 
 16  land_area         2772 non-null   object 


Count the distinct links in all the 2772 links.

In [8]:
df.link.unique().shape

(2772,)

Find out some strange data to delete from dataset

In [9]:
df.describe().round(2)

rental_value  useful_area  m2_value    rooms   suites  parking  \
count  2772.00       2772.00      2772.00   2772.00  2772.00  2772.00   
mean   4523.01       69.16        38.71     1.90     0.61     0.81      
std    46877.06      53.51        25.28     0.93     0.84     0.80      
min    300.00        0.01         1.10      0.00     0.00     0.00      
25%    1100.00       35.00        25.00     1.00     0.00     0.00      
50%    1800.00       55.00        35.00     2.00     0.00     1.00      
75%    3250.00       79.00        46.00     3.00     1.00     1.00      
max    2250000.00    621.00       750.00    5.00     5.00     5.00      

       condo_value  total_area      iptu    floor  
count  2772.00      2772.00     2772.00   2772.00  
mean   520.92       113.92      358.14    1.01     
std    816.06       3799.23     859.39    2.79     
min    0.00         0.00        0.00      0.00     
25%    100.00       0.00        0.00      0.00     
50%    394.00       20.50       31.50     0.00     
75%    662.50       60.00       290.25    0.00     
max    29159.00     200000.00   10633.00  31.00

Dropping properties registered with rental value upper than R$ 10.000,00 (maybe due to digitation error)

In [10]:
drop_rental_value = df.loc[df.rental_value >= 10000]
drop_rental_value

link  \
12    https://www.dfimoveis.com.br/imovel/apartamento-4-quartos-aluguel-noroeste-brasilia-df-sqnw-110-714976                    
18    https://www.dfimoveis.com.br/imovel/apartamento-3-quartos-aluguel-asa-norte-brasilia-df-condominio-ilhas-do-lago-750772   
24    https://www.dfimoveis.com.br/imovel/apartamento-3-quartos-aluguel-park-sul-brasilia-df-sgcv-lote-27-775437                
51    https://www.dfimoveis.com.br/imovel/apartamento-3-quartos-aluguel-sudoeste-brasilia-df-sqsw-306-775454                    
80    https://www.dfimoveis.com.br/imovel/apartamento-4-quartos-aluguel-taguatinga-norte-taguatinga-df-cnb-9-778951             
...                                                                                                             ...             
2623  https://www.dfimoveis.com.br/imovel/apartamento-4-quartos-aluguel-asa-norte-brasilia-df-sqn-314-bloco-h-703909            
2666  https://www.dfimoveis.com.br/imovel/apartamento-4-quartos-aluguel-noroeste-brasilia-df-sqnw-108-778488                    
2676  https://www.dfimoveis.com.br/imovel/apartamento-3-quartos-aluguel-park-sul-brasilia-df-sgcv-lote-27-258983                
2677  https://www.dfimoveis.com.br/imovel/apartamento-4-quartos-aluguel-sudoeste-brasilia-df-sqsw-300-bloco-r-733645            
2720  https://www.dfimoveis.com.br/imovel/apartamento-3-quartos-aluguel-asa-sul-brasilia-df-sqs-207-bloco-f-507992              

                                         title          dt_scraping  \
12    SQNW 110                                  2023-06-05 22:17:13   
18    Condomínio Ilhas do Lago                  2023-06-05 22:17:15   
24    SGCV Lote 27  - Prime Residence           2023-06-05 22:17:15   
51    SQSW 306                                  2023-06-05 22:17:17   
80    CNB 9                                     2023-06-05 22:17:19   
...     ...                                                     ...   
2623  SQN 314 Bloco H                           2023-06-05 22:18:52   
2666  SQNW 108                                  2023-06-05 22:18:53   
2676  SGCV Lote 27  - Park Sul Prime Residence  2023-06-05 22:18:54   
2677  SQSW 300 Bloco R                          2023-06-05 22:18:54   
2720  SQS 207 Bloco F                           2023-06-05 22:18:55   

      rental_value  useful_area  m2_value      neighborhood        city  \
12    11800.0       192.00       61.0      NOROESTE          BRASÍLIA     
18    12000.0       94.00        127.0     ASA NORTE         BRASÍLIA     
24    12000.0       147.44       81.0      PARK SUL          BRASÍLIA     
51    16000.0       349.00       45.0      SUDOESTE          BRASÍLIA     
80    10000.0       350.00       28.0      TAGUATINGA NORTE  TAGUATINGA   
...       ...          ...        ...                   ...         ...   
2623  13000.0       320.00       40.0      ASA NORTE         BRASÍLIA     
2666  22000.0       403.00       54.0      NOROESTE          BRASÍLIA     
2676  11000.0       242.00       45.0      PARK SUL          BRASÍLIA     
2677  15800.0       310.00       50.0      SUDOESTE          BRASÍLIA     
2720  10190.0       175.00       58.0      ASA SUL           BRASÍLIA     

      rooms  suites  parking  condo_value  total_area    iptu  floor  \
12    4      4       2        1815.0       0.00        1027.0  0       
18    3      1       2        2144.0       0.00        139.0   3       
24    3      3       2        1450.0       147.44      0.0     9       
51    3      3       4        2643.0       349.00      6386.0  0       
80    5      4       4        2500.0       350.00      0.0     0       
...  ..     ..      ..           ...          ...      ...    ..       
2623  4      2       4        3900.0       320.00      808.0   0       
2666  4      4       5        2982.0       0.00        0.0     0       
2676  3      3       2        2033.0       0.00        252.0   9       
2677  4      4       0        3900.0       0.00        1066.0  0       
2720  3      1       1    

In [11]:
df.drop(drop_rental_value.index, inplace=True)
df.describe().round(2)

rental_value  useful_area  m2_value    rooms   suites  parking  \
count  2657.00       2657.00      2657.00   2657.00  2657.00  2657.00   
mean   2359.84       63.06        37.70     1.83     0.51     0.73      
std    1844.21       41.41        24.77     0.87     0.68     0.68      
min    300.00        0.01         1.10      0.00     0.00     0.00      
25%    1100.00       35.00        24.00     1.00     0.00     0.00      
50%    1700.00       52.00        34.00     2.00     0.00     1.00      
75%    3000.00       74.00        44.00     2.00     1.00     1.00      
max    9900.00       468.00       750.00    5.00     5.00     4.00      

       condo_value  total_area     iptu   floor  
count  2657.00      2657.00     2657.00  2657.0  
mean   465.95       114.20      317.33   1.0     
std    740.01       3880.51     722.13   2.8     
min    0.00         0.00        0.00     0.0     
25%    90.00        0.00        0.00     0.0     
50%    385.00       25.00       35.00    0.0     
75%    620.00       57.00       266.00   0.0     
max    29159.00     200000.00   9218.00  31.0

Deleting data with condo values upper then 3000

In [12]:
drop_condo_value = df.loc[df.condo_value >= 3000]
drop_condo_value

link  \
21    https://www.dfimoveis.com.br/imovel/apartamento-4-quartos-aluguel-sudoeste-brasilia-df-sqsw-305-bloco-j-752555                          
834   https://www.dfimoveis.com.br/imovel/apartamento-4-quartos-aluguel-sudoeste-brasilia-df-sqsw-305-bloco-j-262639                          
1095  https://www.dfimoveis.com.br/imovel/kitnet-1-quarto-aluguel-sudoeste-brasilia-df-qmsw-5-lote-2-bloco-b-771896                           
1566  https://www.dfimoveis.com.br/imovel/apartamento-1-quarto-aluguel-setor-placa-da-mercedes-nucleo-bandeirante-df-splm-conjunto-6-774992   
2759  https://www.dfimoveis.com.br/imovel/apartamento-4-quartos-aluguel-sudoeste-brasilia-df-sqsw-305-bloco-j-751432                          

                                         title          dt_scraping  \
21    SQSW 305 Bloco J                          2023-06-05 22:17:15   
834   SQSW 305 Bloco J  - Via Parque da Cidade  2023-06-05 22:17:49   
1095  QMSW 5 Lote 2 Bloco B                     2023-06-05 22:17:58   
1566  SPLM Conjunto 6                           2023-06-05 22:18:14   
2759  SQSW 305 Bloco J                          2023-06-05 22:18:56   

      rental_value  useful_area  m2_value             neighborhood  \
21    8999.0        201.0        44.0      SUDOESTE                  
834   9300.0        200.0        46.0      SUDOESTE                  
1095  950.0         27.0         35.0      SUDOESTE                  
1566  650.0         35.0         18.0      SETOR PLACA DA MERCEDES   
2759  9300.0        200.0        46.0      SUDOESTE                  

                    city  rooms  suites  parking  condo_value  total_area  \
21    BRASÍLIA            4      4       3        3250.0       263.0        
834   BRASÍLIA            4      4       3        3325.0       200.0        
1095  BRASÍLIA            1      1       1        29159.0      32.0         
1566  NÚCLEO BANDEIRANTE  1      0       0        12000.0      0.0          
2759  BRASÍLIA            4      4       3        3325.0       200.0        

        iptu  floor floor_properties land_area financing exchange agio  
21    416.0   1      2                                                  
834   4988.0  1                                                         
1095  805.0   0                                                         
1566  0.0     0                                                         
2759  4988.0  1

In [13]:
df.drop(drop_condo_value.index, inplace=True)
df.describe().round(2)

rental_value  useful_area  m2_value    rooms   suites  parking  \
count  2652.00       2652.00      2652.00   2652.00  2652.00  2652.00   
mean   2353.28       62.93        37.70     1.83     0.51     0.73      
std    1831.03       41.18        24.78     0.86     0.68     0.68      
min    300.00        0.01         1.10      0.00     0.00     0.00      
25%    1100.00       35.00        24.00     1.00     0.00     0.00      
50%    1700.00       52.00        34.00     2.00     0.00     1.00      
75%    2990.00       74.00        44.00     2.00     1.00     1.00      
max    9900.00       468.00       750.00    5.00     5.00     4.00      

       condo_value  total_area     iptu    floor  
count  2652.00      2652.00     2652.00  2652.00  
mean   447.57       114.15      313.70   1.00     
std    422.47       3884.16     711.24   2.81     
min    0.00         0.00        0.00     0.00     
25%    88.75        0.00        0.00     0.00     
50%    383.00       25.00       34.00    0.00     
75%    617.25       57.00       265.00   0.00     
max    2375.00      200000.00   9218.00  31.00

Deleting data with total area upper than 5000 m²

In [14]:
drop_total_area = df.loc[df.total_area >= 5000]
drop_total_area

link  \
1323  https://www.dfimoveis.com.br/imovel/apartamento-1-quarto-aluguel-asa-sul-brasilia-df-shs-quadra-6-738629   

             title          dt_scraping  rental_value  useful_area  m2_value  \
1323  SHS Quadra 6  2023-06-05 22:18:06  5500.0        60.0         91.0       

     neighborhood      city  rooms  suites  parking  condo_value  total_area  \
1323  ASA SUL      BRASÍLIA  1      0       0        2280.0       200000.0     

      iptu  floor floor_properties land_area financing exchange agio  
1323  0.0   0

In [15]:
df.drop(drop_total_area.index, inplace=True)
df.describe().round(2)

rental_value  useful_area  m2_value    rooms   suites  parking  \
count  2651.00       2651.00      2651.00   2651.00  2651.00  2651.00   
mean   2352.09       62.93        37.68     1.83     0.51     0.73      
std    1830.35       41.19        24.77     0.86     0.68     0.68      
min    300.00        0.01         1.10      0.00     0.00     0.00      
25%    1100.00       35.00        24.00     1.00     0.00     0.00      
50%    1700.00       52.00        34.00     2.00     0.00     1.00      
75%    2990.00       74.00        44.00     2.00     1.00     1.00      
max    9900.00       468.00       750.00    5.00     5.00     4.00      

       condo_value  total_area     iptu    floor  
count  2651.00      2651.00     2651.00  2651.00  
mean   446.88       38.75       313.82   1.00     
std    421.05       97.95       711.35   2.81     
min    0.00         0.00        0.00     0.00     
25%    87.50        0.00        0.00     0.00     
50%    383.00       25.00       34.00    0.00     
75%    617.00       56.86       265.00   0.00     
max    2375.00      3000.00     9218.00  31.00

End cleaning, reset index and save to a CSV file

In [16]:
df.reset_index(inplace=True, drop=True)
df.to_csv(output_file, index=False)

In [17]:
csv = pd.read_csv(output_file)
csv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2651 entries, 0 to 2650
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   link              2651 non-null   object 
 1   title             2651 non-null   object 
 2   dt_scraping       2651 non-null   object 
 3   rental_value      2651 non-null   float64
 4   useful_area       2651 non-null   float64
 5   m2_value          2651 non-null   float64
 6   neighborhood      2651 non-null   object 
 7   city              2651 non-null   object 
 8   rooms             2651 non-null   int64  
 9   suites            2651 non-null   int64  
 10  parking           2651 non-null   int64  
 11  condo_value       2651 non-null   float64
 12  total_area        2651 non-null   float64
 13  iptu              2651 non-null   float64
 14  floor             2651 non-null   int64  
 15  floor_properties  204 non-null    float64
 16  land_area         43 non-null     object 
